source/opt/scalar_analysis_nodes.h

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASI,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_SCALAR_ANALYSIS_NODES_H_
+#define LIBSPIRV_OPT_SCALAR_ANALYSIS_NODES_H_
+
+#include "tree_iterator.h"
+
+namespace spvtools {
+namespace opt {
+
+class ScalarEvolutionAnalysis;
+class SEConstantNode;
+class SERecurrentNode;
+class SEAddNode;
+class SEMultiplyNode;
+class SENegative;
+class SEValueUnknown;
+class SECantCompute;
+
+// Abstract class representing a node in the scalar evolution DAG. Each node
+// contains a vector of pointers to its children and each subclass of SENode
+// implements GetType and an As method to allow casting. SENodes can be hashed
+// using the SENodeHash functor. The vector of children is sorted when a node is
+// added. This is important as it allows the hash of X+Y to be the same as Y+X.
+class SENode {
+ public:
+  enum SENodeType {
+    Constant,
+    RecurrentExpr,
+    Add,
+    Multiply,
+    Negative,
+    ValueUnknown,
+    CanNotCompute
+  };
+
+  explicit SENode(opt::ScalarEvolutionAnalysis* parent_analysis)
+      : parent_analysis_(parent_analysis) {}
+
+  virtual SENodeType GetType() const = 0;
+
+  virtual ~SENode() {}
+
+  inline void AddChild(SENode* child) {
+    children_.push_back(child);
+
+    // Children are sorted so the hashing
```

Comment:
Not sure why we need sorting.  But, if we do, perhaps sorting by result or unique ID will be better.  We should not be sorting pointers.

Reply:
Nodes are now assigned an ID on creation and sorted by that. We sort so that nodes with the same children added in a different order have the same hash and are easier to compare @dnovillo . So X+Y+Z should be the same as Z+X+Y